### ISSUE TITLE
Proposal: extending TaskEither with fold

### ISSUE THREAD

**Issue by pixeleet** [nullpub/fun#32]
Using TaskEither a lot and currently that means writing a lot of the following code:
```ts
pipe(task_either, t => t(), then(E.fold(err, ok)))
```

This could be easily simplified to:
```ts
pipe(task_either, TE.fold(err, ok))
```

Proposed fold impl:
```ts
export function fold<L, R, B>(
  onLeft: (left: L) => B,
  onRight: (right: R) => B
): (ta: TaskEither<L, R>) => Task<B> {
  return (ta) =>
    pipe(
      ta(),
      then(eitherFold<L, R, B>(onLeft, onRight)),
      constant
    );
}
```

**Comment by baetheus** [comment]
This looks good but the early application of `ta()` drops the laziness of Task by doing an immediate execution of the TaskEither. It would probably be better as:

```ts
export function fold<L, R, B>(
  onLeft: (left: L) => B,
  onRight: (right: R) => B
): (ta: TaskEither<L, R>) => Task<B> {
  return (ta) => async () => {
    const result = await ta();
    return eitherFold(result);
  }
}
```

**Comment by pixeleet** [comment]
Yes, good catch! 

How about this?
```ts
export function fold<L, R, B>(
  onLeft: (left: L) => B,
  onRight: (right: R) => B
): (ta: TaskEither<L, R>) => Task<B> {
  return (ta) => async () => {
    return await pipe(ta(), then(eitherFold(onLeft, onRight)))
  }
}
```

Or there's a version that compiles correctly without explicit async too
```ts
export function fold<L, R, B>(
  onLeft: (left: L) => B,
  onRight: (right: R) => B
): (ta: TaskEither<L, R>) => Task<B> {
  return (ta) => () => 
    pipe(ta(), then(eitherFold(onLeft, onRight)))
}
```

Or maybe

```ts
export function fold<L, R, B>(
  onLeft: (left: L) => B,
  onRight: (right: R) => B
): (ta: TaskEither<L, R>) => Task<B> {
  return (ta) =>
    pipe(ta(), then(eitherFold(onLeft, onRight)), constant)
}
```

Alternatively, I was lately thinking of introducing: 
```ts
/**
 * Run a function
 */
export function run<A>(fa: () => A): A {
  return fa();
}
```

Which in turn would let us do cool stuff like using flow (to create a pointfree version): 
```ts
export function fold<L, R, B>(
  onLeft: (left: L) => B,
  onRight: (right: R) => B,
): (ta: TaskEither<L, R>) => Task<B> {
  return flow(run, then(eitherFold(onLeft, onRight)), constant);
}
```

**Comment by baetheus** [comment]
Let's avoid adding `pipe` and `then` here as they don't really add to readability or performance. In fact, they both introduce thunks that might be optimized by some browser vms.

**Comment by pixeleet** [comment]
So then I guess the best way to write this would be:
```ts
export function fold<L, R, B>(
  onLeft: (left: L) => B,
  onRight: (right: R) => B,
): (ta: TaskEither<L, R>) => Task<B> {
  return (ta) =>
    () =>
      ta().then(eitherFold<L, R, B>(onLeft, onRight)).catch(onLeft);
}
```

**Comment by baetheus** [comment]
This looks good but we should avoid using catch here as the contract of TaskEither is that the promise never throws.

**Comment by pixeleet** [comment]
Similarly for Affect I'm seeing a use-case of implementing fold as I'm starting to use it more and more. 
Something along the lines of
```ts
export function fold<C, L, R, B>(
  onLeft: (left: L) => B,
  onRight: (right: R) => B,
): (ta: Affect<C, L, R>) => Reader<C, Promise<B>> {
  return (ta) => (c) => ta(c).then(eitherFold(onLeft, onRight));
}
```

btw: I'm already using these in apps from my fork (v2) branch.

**Comment by baetheus** [comment]
I've come around on adding fold for Affect, TaskEither, and whatever other ADTs we can find that could use it. Let's get these sorted and landed ASAP.

**Comment by pixeleet** [comment]
Which ADTs should we go for? 
We agree on TaskEither, Affect, what other types would you find it useful to implement for?

**Comment by baetheus** [comment]
Fixed in #35
